feat: Indicate repeated tag

magicFeirl opened the pull request:
Hello! This PR adds a 🔁 icon to an autocomplete item when a tag exists among previous prompt tags. This feature works only if the tagfile is Danbooru or e621.

![ezgif-1-0e8bb7acd3](https://github.com/user-attachments/assets/cccf4eea-a0e0-4dc4-8a27-51fa65945e99)

DominikDoom commented on javascript/tagAutocomplete.js:
```diff
@@ -831,6 +858,15 @@ function addResultsToList(textArea, results, tagword, resetList) {
+            const span = document.createElement("span")
+            textNode.insertBefore(span, textNode.firstChild)
+            span.classList.add("existed")
+            span.title = `🔁 This tag has existed in before prompt.`
```
This would be better grammatically, also no need for ` as it's not interpolated
```suggestion
            span.title = "🔁 The prompt already contains this tag";
```

DominikDoom commented on javascript/tagAutocomplete.js:
```diff
@@ -90,6 +90,10 @@ const autocompleteCSS = `
         content: "✨";
         margin-right: 2px;
     }
+    .acMetaText span.existed::after {
```
Just a name suggestion, existed sounds like it isn't there anymore
```suggestion
    .acMetaText span.used::after {
```

DominikDoom commented on javascript/tagAutocomplete.js:
```diff
@@ -831,6 +858,15 @@ function addResultsToList(textArea, results, tagword, resetList) {
+            const textNode = flexDiv.querySelector(".acMetaText")
+            const span = document.createElement("span")
+            textNode.insertBefore(span, textNode.firstChild)
+            span.classList.add("existed")
```
Same as in the style, class name suggestion
```suggestion
            span.classList.add("used");
```

DominikDoom commented on javascript/tagAutocomplete.js:
```diff
@@ -666,6 +670,30 @@ function addResultsToList(textArea, results, tagword, resetList) {
     let nextLength = Math.min(results.length, resultCount + TAC_CFG.resultStepLength);
+    const IS_DAN_OR_E621_TAG_FILE =  (tagFileName.toLowerCase().startsWith("danbooru") || tagFileName.toLowerCase().startsWith("e621"))
+
+    let prompt = "", tagCount = {}
+
+    // Indicate if has existed tag before 
+    if(IS_DAN_OR_E621_TAG_FILE) {
+        prompt = textArea.value.trim()
```
Not a fan of this double variable initialization. I'd suggest moving `prompt` into the if scope below as you only use it there anyway.

```suggestion
    const tagCount = {};

    // Indicate if tag was used before 
    if (IS_DAN_OR_E621_TAG_FILE) {
        const prompt = textArea.value.trim();
```

DominikDoom commented on javascript/tagAutocomplete.js:
```diff
@@ -666,6 +670,30 @@ function addResultsToList(textArea, results, tagword, resetList) {
+        }).map(tag => tag.replaceAll(" ", "_").replaceAll("\\(", "(").replaceAll("\\)", ")"))
+    
+        // Split tags by `,`  and count tag 
+        for(const tag of unsanitizedTags) {
```
```suggestion
        for (const tag of unsanitizedTags) {
```

DominikDoom commented on javascript/tagAutocomplete.js:
```diff
@@ -831,6 +858,15 @@ function addResultsToList(textArea, results, tagword, resetList) {
             flexDiv.title = "✨ Frequent tag. Ctrl/Cmd + click to reset usage count."
         }
 
+        // Add 🔁 to indicate if has existed tag
+        if(IS_DAN_OR_E621_TAG_FILE && tagCount[result.text] >= 1) {
```
```suggestion
        // Add 🔁 to indicate if tag was used before
        if (IS_DAN_OR_E621_TAG_FILE && tagCount[result.text] >= 1) {
```

DominikDoom reviewed (changes requested):
First of all, good idea 👍 

I have some change suggestions, as detailed in the review comments.
Additionally, please follow the style of the rest of the code and add `;` at the end of lines (everywhere in your code, not just where I suggested changes). Yes, it's optional, but prevents errors. Also spaces after `if` and `for` like I marked. I don't have an auto-formatter running before merges, so please fix it manually.

Also, a general question:
What's the reason to only apply this for danbooru / e621 tag files?
It should work with all tag files, or did I miss something?

magicFeirl commented:
> I have some change suggestions, as detailed in the review comments. Additionally, please follow the style of the rest of the code and add `;` at the end of lines (everywhere in your code, not just where I suggested changes). Yes, it's optional, but prevents errors. Also spaces after `if` and `for` like I marked. I don't have an auto-formatter running before merges, so please fix it manually.

Thanks! I have fixed the code style.
 
> Also, a general question: What's the reason to only apply this for danbooru / e621 tag files? It should work with all tag files, or did I miss something?

I only use the Danbooru-style tags, which are separated by `,`, and I am not sure if other tag styles could work, so I added the filename filter.

DominikDoom commented:
One more thing: I tried it out and found a small bug here:
![image](https://github.com/user-attachments/assets/8e2ff602-8d0a-474f-b965-daec536d216e)

Due to the suggestion exactly matching what I typed, it was shown as repeated.
While that's technically correct, I feel it would be better to not apply in this case as I haven't actually chosen a suggestion yet.

I'm not 100% sure, but I think to fix this, you could compare `result.text` with `tagword` and don't add the marker if they are equal and tagCount is exactly 1 (meaning the one we just typed is the only one).

magicFeirl commented:
Thanks for suggestion. I have fixed this issue.

![fix](https://github.com/user-attachments/assets/9ec84baa-db20-47bd-bc1b-4d4fdc09eecd)
